internal/controller/dpuoperatorconfig_controller_test.go

```diff
@@ -0,0 +1,172 @@
+
+		By("verify test DPU operator namespace is created")
+		retrieved_namespace := &corev1.Namespace{}
+		err = k8sClient.Get(context.Background(), client.ObjectKey{Name: testNamespace}, retrieved_namespace)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(retrieved_namespace.ObjectMeta.Name).To(Equal(testNamespace))
+
+		By("setup controller manager")
+		k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+			Scheme: scheme.Scheme,
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		By("setup controller manager reconciler")
+		err = (&DpuOperatorConfigReconciler{
+			Client: k8sManager.GetClient(),
+			Scheme: k8sManager.GetScheme(),
+		}).SetupWithManager(k8sManager)
+		Expect(err).ToNot(HaveOccurred())
+
+		ctx, cancel = context.WithCancel(context.Background())
+
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer GinkgoRecover()
+			By("start controller manager")
+			err := k8sManager.Start(ctx)
+			Expect(err).ToNot(HaveOccurred())
+		}()
+
+		DeferCleanup(func() {
+			By("shut down controller manager")
+			cancel()
+			wg.Wait()
+		})
+
+		// HACK: Wait until controller manager boots up
+		// FIXME: Find a better way to detect that the controller manager is in a good state. Perhaps we will implement
+		// the "status" portion of the CR in the future.
+		time.Sleep(3 * time.Second)
+	})
+
+	Context("When controller manager has started without DpuOperatorConfig CRD", func() {
+		BeforeAll(func() {
+			By("create DpuOperatorConfig CRD")
+			config := &configv1.DpuOperatorConfig{}
+			config.SetNamespace(testNamespace)
+			config.SetName(testDpuOperatorConfigName)
+			config.Spec = configv1.DpuOperatorConfigSpec{
+				Mode:     "host",
+				LogLevel: 2,
+			}
+			Expect(k8sClient.Create(context.Background(), config)).Should(Succeed())
+			DeferCleanup(func() {
+				err := k8sClient.Delete(context.Background(), config)
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			By("verify DpuOperatorConfig CRD is created")
+			retrieved_config := &configv1.DpuOperatorConfig{}
+			err := k8sClient.Get(context.Background(), client.ObjectKey{Namespace: testNamespace, Name: testDpuOperatorConfigName}, retrieved_config)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(retrieved_config.ObjectMeta.Namespace).To(Equal(testNamespace))
+			Expect(retrieved_config.ObjectMeta.Name).To(Equal(testDpuOperatorConfigName))
+		})
+		It("should have DPU daemon daemonsets created by controller manager", func() {
+			daemonSet := &appsv1.DaemonSet{}
+			err := WaitForDaemonSetReady(daemonSet, k8sClient, testNamespace, testDpuDaemonName)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(daemonSet.OwnerReferences).To(HaveLen(1))
+			Expect(daemonSet.OwnerReferences[0].Kind).To(Equal(testDpuOperatorConfigKind))
+			Expect(daemonSet.OwnerReferences[0].Name).To(Equal(testDpuOperatorConfigName))
+		})
+		It("change mode to \"dpu\" in DpuOperatorConfig CRD to make sure reconcile() loop detects new changes", func() {
```

Comment:
not implemented. I would not test it because it is confusing that there is a test for something we didn't implement properly.

Reply:
Done. Let's seperate the tests as discussed between DPU and Host.